File: frontend/src/models/linking.ts
import { dto } from "@/wailsjs/wailsjs/go/models";

// =========================================================================
// Linking Types
// =========================================================================

export type PlanStatus = "draft" | "suggesting" | "ready" | "applying" | "applied" | "failed";
export type LinkStatus = "planned" | "approved" | "rejected" | "applying" | "applied" | "failed";
export type LinkSource = "ai" | "manual";

export interface LinkPlan {
    id: number;
    sitemapId: number;
    siteId: number;
    name: string;
    status: PlanStatus;
    providerId?: number;
    promptId?: number;
    error?: string;
    createdAt: string;
    updatedAt: string;
}

export interface PlannedLink {
    id: number;
    planId: number;
    sourceNodeId: number;
    targetNodeId: number;
    anchorText?: string;
    anchorContext?: string;
    status: LinkStatus;
    source: LinkSource;
    position?: number;
    confidence?: number;
    error?: string;
    appliedAt?: string;
    createdAt: string;
    updatedAt: string;
}

export interface GraphNode {
    nodeId: number;
    title: string;
    slug: string;
    path: string;
    hasContent: boolean;
    outgoingLinkCount: number;
    incomingLinkCount: number;
}

export interface GraphEdge {
    id: number;
    sourceNodeId: number;
    targetNodeId: number;
    anchorText?: string;
    status: LinkStatus;
    source: LinkSource;
    confidence?: number;
}

export interface LinkGraph {
    nodes: GraphNode[];
    edges: GraphEdge[];
}

// =========================================================================
// Input Types
// =========================================================================

export interface CreateLinkPlanInput {
    sitemapId: number;
    siteId: number;
    name: string;
}

export interface AddLinkInput {
    planId: number;
    sourceNodeId: number;
    targetNodeId: number;
}

export interface UpdateLinkInput {
    id: number;
    anchorText?: string;
    anchorContext?: string;
}

export interface SuggestLinksInput {
    planId: number;
    providerId: number;
    promptId?: number;
    nodeIds?: number[];
    feedback?: string;
    maxIncoming?: number;
    maxOutgoing?: number;
}

export interface ApplyLinksInput {
    planId: number;
    linkIds: number[];
    providerId: number;
    promptId: number;
}

export interface ApplyLinksResult {
    totalLinks: number;
    appliedLinks: number;
    failedLinks: number;
}

// =========================================================================
// Apply Links Event Types (PascalCase - as sent from Go)
// =========================================================================

export interface ApplyStartedEvent {
    TaskID: string;
    TotalLinks: number;
    TotalPages: number;
}

export interface ApplyProgressEvent {
    TaskID: string;
    ProcessedPages: number;
    TotalPages: number;
    AppliedLinks: number;
    FailedLinks: number;
    CurrentPage?: PageInfo;
}

export interface PageInfo {
    NodeID: number;
    Title: string;
    Path: string;
}

export interface ApplyCompletedEvent {
    TaskID: string;
    TotalLinks: number;
    AppliedLinks: number;
    FailedLinks: number;
    DurationMs: number;
}

export interface ApplyFailedEvent {
    TaskID: string;
    Error: string;
}

export interface PageProcessingEvent {
    TaskID: string;
    NodeID: number;
    Title: string;
    LinkCount: number;
}

export interface PageCompletedEvent {
    TaskID: string;
    NodeID: number;
    Title: string;
    AppliedLinks: number;
    FailedLinks: number;
}

export interface PageFailedEvent {
    TaskID: string;
    NodeID: number;
    Title: string;
    Error: string;
}

// =========================================================================
// Suggest Links Event Types (PascalCase - as sent from Go)
// =========================================================================

export interface SuggestStartedEvent {
    TaskID: string;
    TotalNodes: number;
    TotalBatches: number;
}

export interface SuggestProgressEvent {
    TaskID: string;
    CurrentBatch: number;
    TotalBatches: number;
    ProcessedNodes: number;
    TotalNodes: number;
    LinksCreated: number;
    CurrentBatchSize: number;
}

export interface SuggestCompletedEvent {
    TaskID: string;
    TotalNodes: number;
    LinksCreated: number;
    DurationMs: number;
}

export interface SuggestFailedEvent {
    TaskID: string;
    Error: string;
}

export interface SuggestCancelledEvent {
    TaskID: string;
    ProcessedNodes: number;
    LinksCreated: number;
}

export interface ApplyCancelledEvent {
    TaskID: string;
    ProcessedPages: number;
    AppliedLinks: number;
}

// =========================================================================
// Mappers
// =========================================================================

export function mapLinkPlan(x: dto.LinkPlan): LinkPlan {
    return {
        id: x.id,
        sitemapId: x.sitemapId,
        siteId: x.siteId,
        name: x.name,
        status: x.status as PlanStatus,
        providerId: x.providerId || undefined,
        promptId: x.promptId || undefined,
        error: x.error || undefined,
        createdAt: x.createdAt,
        updatedAt: x.updatedAt,
    };
}

export function mapPlannedLink(x: dto.PlannedLink): PlannedLink {
    return {
        id: x.id,
        planId: x.planId,
        sourceNodeId: x.sourceNodeId,
        targetNodeId: x.targetNodeId,
        anchorText: x.anchorText || undefined,
        anchorContext: x.anchorContext || undefined,
        status: x.status as LinkStatus,
        source: x.source as LinkSource,
        position: x.position || undefined,
        confidence: x.confidence || undefined,
        error: x.error || undefined,
        appliedAt: x.appliedAt || undefined,
        createdAt: x.createdAt,
        updatedAt: x.updatedAt,
    };
}

export function mapGraphNode(x: dto.GraphNode): GraphNode {
    return {
        nodeId: x.nodeId,
        title: x.title,
        slug: x.slug,
        path: x.path,
        hasContent: x.hasContent,
        outgoingLinkCount: x.outgoingLinkCount,
        incomingLinkCount: x.incomingLinkCount,
    };
}

export function mapGraphEdge(x: dto.GraphEdge): GraphEdge {
    return {
        id: x.id,
        sourceNodeId: x.sourceNodeId,
        targetNodeId: x.targetNodeId,
        anchorText: x.anchorText || undefined,
        status: x.status as LinkStatus,
        source: x.source as LinkSource,
        confidence: x.confidence || undefined,
    };
}

export function mapLinkGraph(x: dto.LinkGraph): LinkGraph {
    return {
        nodes: (x.nodes || []).map(mapGraphNode),
        edges: (x.edges || []).map(mapGraphEdge),
    };
}

File: internal/dto/linking.go
package dto

import (
	"time"

	"github.com/davidmovas/postulator/internal/domain/linking"
)

// LinkPlan DTOs

type LinkPlan struct {
	ID         int64     `json:"id"`
	SitemapID  int64     `json:"sitemapId"`
	SiteID     int64     `json:"siteId"`
	Name       string    `json:"name"`
	Status     string    `json:"status"`
	ProviderID *int64    `json:"providerId,omitempty"`
	PromptID   *int64    `json:"promptId,omitempty"`
	Error      *string   `json:"error,omitempty"`
	CreatedAt  time.Time `json:"createdAt"`
	UpdatedAt  time.Time `json:"updatedAt"`
}

func NewLinkPlan(p *linking.LinkPlan) *LinkPlan {
	return &LinkPlan{
		ID:         p.ID,
		SitemapID:  p.SitemapID,
		SiteID:     p.SiteID,
		Name:       p.Name,
		Status:     string(p.Status),
		ProviderID: p.ProviderID,
		PromptID:   p.PromptID,
		Error:      p.Error,
		CreatedAt:  p.CreatedAt,
		UpdatedAt:  p.UpdatedAt,
	}
}

func NewLinkPlans(plans []*linking.LinkPlan) []*LinkPlan {
	result := make([]*LinkPlan, len(plans))
	for i, p := range plans {
		result[i] = NewLinkPlan(p)
	}
	return result
}

// PlannedLink DTOs

type PlannedLink struct {
	ID            int64      `json:"id"`
	PlanID        int64      `json:"planId"`
	SourceNodeID  int64      `json:"sourceNodeId"`
	TargetNodeID  int64      `json:"targetNodeId"`
	AnchorText    *string    `json:"anchorText,omitempty"`
	AnchorContext *string    `json:"anchorContext,omitempty"`
	Status        string     `json:"status"`
	Source        string     `json:"source"`
	Position      *int       `json:"position,omitempty"`
	Confidence    *float64   `json:"confidence,omitempty"`
	Error         *string    `json:"error,omitempty"`
	AppliedAt     *time.Time `json:"appliedAt,omitempty"`
	CreatedAt     time.Time  `json:"createdAt"`
	UpdatedAt     time.Time  `json:"updatedAt"`
}

func NewPlannedLink(l *linking.PlannedLink) *PlannedLink {
	return &PlannedLink{
		ID:            l.ID,
		PlanID:        l.PlanID,
		SourceNodeID:  l.SourceNodeID,
		TargetNodeID:  l.TargetNodeID,
		AnchorText:    l.AnchorText,
		AnchorContext: l.AnchorContext,
		Status:        string(l.Status),
		Source:        string(l.Source),
		Position:      l.Position,
		Confidence:    l.Confidence,
		Error:         l.Error,
		AppliedAt:     l.AppliedAt,
		CreatedAt:     l.CreatedAt,
		UpdatedAt:     l.UpdatedAt,
	}
}

func NewPlannedLinks(links []*linking.PlannedLink) []*PlannedLink {
	result := make([]*PlannedLink, len(links))
	for i, l := range links {
		result[i] = NewPlannedLink(l)
	}
	return result
}

// LinkGraph DTOs

type LinkGraph struct {
	Nodes []*GraphNode `json:"nodes"`
	Edges []*GraphEdge `json:"edges"`
}

type GraphNode struct {
	NodeID            int64  `json:"nodeId"`
	Title             string `json:"title"`
	Slug              string `json:"slug"`
	Path              string `json:"path"`
	HasContent        bool   `json:"hasContent"`
	OutgoingLinkCount int    `json:"outgoingLinkCount"`
	IncomingLinkCount int    `json:"incomingLinkCount"`
}

type GraphEdge struct {
	ID           int64    `json:"id"`
	SourceNodeID int64    `json:"sourceNodeId"`
	TargetNodeID int64    `json:"targetNodeId"`
	AnchorText   *string  `json:"anchorText,omitempty"`
	Status       string   `json:"status"`
	Source       string   `json:"source"`
	Confidence   *float64 `json:"confidence,omitempty"`
}

func NewLinkGraph(g *linking.LinkGraph) *LinkGraph {
	nodes := make([]*GraphNode, len(g.Nodes))
	for i, n := range g.Nodes {
		nodes[i] = &GraphNode{
			NodeID:            n.NodeID,
			Title:             n.Title,
			Slug:              n.Slug,
			Path:              n.Path,
			HasContent:        n.HasContent,
			OutgoingLinkCount: n.OutgoingLinkCount,
			IncomingLinkCount: n.IncomingLinkCount,
		}
	}

	edges := make([]*GraphEdge, len(g.Edges))
	for i, e := range g.Edges {
		edges[i] = &GraphEdge{
			ID:           e.ID,
			SourceNodeID: e.SourceNodeID,
			TargetNodeID: e.TargetNodeID,
			AnchorText:   e.AnchorText,
			Status:       string(e.Status),
			Source:       string(e.Source),
			Confidence:   e.Confidence,
		}
	}

	return &LinkGraph{
		Nodes: nodes,
		Edges: edges,
	}
}

// Request DTOs

type CreateLinkPlanRequest struct {
	SitemapID int64  `json:"sitemapId"`
	SiteID    int64  `json:"siteId"`
	Name      string `json:"name"`
}

type AddLinkRequest struct {
	PlanID       int64 `json:"planId"`
	SourceNodeID int64 `json:"sourceNodeId"`
	TargetNodeID int64 `json:"targetNodeId"`
}

type UpdateLinkRequest struct {
	ID            int64   `json:"id"`
	AnchorText    *string `json:"anchorText,omitempty"`
	AnchorContext *string `json:"anchorContext,omitempty"`
}

type SuggestLinksRequest struct {
	PlanID     int64   `json:"planId"`
	ProviderID int64   `json:"providerId"`
	PromptID   *int64  `json:"promptId,omitempty"`
	NodeIDs    []int64 `json:"nodeIds,omitempty"`
	Feedback   string  `json:"feedback,omitempty"`
}

type ApplyLinksRequest struct {
	PlanID  int64   `json:"planId"`
	LinkIDs []int64 `json:"linkIds"`
}
